File: teste.js
const net = require('net');
const readline = require('readline');

const messageHistory = [];

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function clearScreen() {
    process.stdout.write('\x1B[2J\x1B[0f');
}

function printMessages() {
    clearScreen();
    messageHistory.forEach(msg => {
        console.log(msg);
    });
}

function parseResponse(response) {
    const parts = response.split('\r\n');
    if (parts.length >= 3 && parts[0] === "SEND") {
        const requestInfo = JSON.parse(parts[1]);
        const payload = JSON.parse(parts[2].replace("Payload: ", ""));
        return { requestInfo, payload };
    } else {
        throw new Error("Formato de resposta desconhecido");
    }
}

function receiveResponse(socket) {
    socket.on('data', data => {
        try {
            const response = data.toString();
            if (response) {
                const { requestInfo, payload } = parseResponse(response);
                messageHistory.push(`Mensagem recebida: ${payload.message}`);
                printMessages();
            }
        } catch (error) {
            console.error(`Erro ao receber mensagem: ${error.message}`);
        }
    });

    socket.on('error', error => {
        console.error(`Erro no socket: ${error.message}`);
    });

    socket.on('close', () => {
        console.log('Conexão fechada pelo servidor.');
        process.exit(0);
    });
}

function sendRequest(socket, userIp, message) {
    const request = "SEND\r\n" +
        JSON.stringify({
            request_id: 11876854719037224982,
            status: "Okay",
            success: true,
            user: 123,
            user_name: userIp,
            channel: "/"
        }) + "\r\n" +
        'Payload: ' + JSON.stringify({
            command: "/message",
            input: message
        }) + "\r\n";
    socket.write(request);
}

function connectToServer(host, port) {
    return new Promise((resolve, reject) => {
        const client = new net.Socket();
        client.connect(port, host, () => {
            resolve(client);
        });
        client.on('error', (err) => {
            reject(err);
        });
    });
}

async function main() {
    const serverIp = await new Promise((resolve) => {
        rl.question("Digite o IP do servidor: ", resolve);
    });
    const userIp = await new Promise((resolve) => {
        rl.question("Digite seu IP: ", resolve);
    });

    const host = serverIp;
    const port = 6969;

    try {
        const socketConnection = await connectToServer(host, port);
        receiveResponse(socketConnection);

        console.log("Digite 'SAIR' para sair do chat.");

        rl.on('line', (input) => {
            if (input.toLowerCase() === 'sair') {
                socketConnection.end();
                rl.close();
            } else {
                messageHistory.push(`Você: ${input}`);
                sendRequest(socketConnection, userIp, input);
                printMessages();
            }
        });
    } catch (error) {
        console.error(`Erro ao conectar ao servidor: ${error.message}`);
    }
}

main();

File: teste.py
import socket
import json
import threading

def send_request(socket, user_ip, message):
    # Modifica a solicitação para incluir o IP do usuário como identificador
    request= "SEND\r\n "
    request+= '{ "request_id": 11876854719037224982, "status": "Okay", "success": true, "user": 123, "user_name": "' + user_ip + '", "channel": "/" }\r\n'
    request+= 'Payload: {"command":"/message","input":"'+ message + '"}\r\n'

    socket.sendall(request.encode())

def receive_response(socket):
    while True:
        response = socket.recv(4096).decode()
        if response:
            print("\nMensagem recebida:", response)

def parse_response(response):
    parts = response.split('\r\n')
    if parts[0] == "RECEIVE":
        request_info = json.loads(parts[1])
        payload = json.loads(parts[2].replace("Payload: ", ""))
        return request_info, payload
    else:
        raise ValueError("Formato de resposta desconhecido")

def main():
    server_ip = input("Digite o IP do servidor: ")
    user_ip = input("Digite seu IP: ")
    host = server_ip
    port = 6969

    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((host, port))

    # Inicia uma thread para receber mensagens do servidor
    threading.Thread(target=receive_response, args=(s,), daemon=True).start()

    print("Digite 'SAIR' para sair do chat.")
    while True:
        message = input("Digite sua mensagem: ")
        if message.lower() == "sair":
            break
        send_request(s, user_ip, message)

    s.close()

if __name__ == "__main__":
    main()